File: sdks/php/src/Edge.php
<?php

declare(strict_types=1);

namespace GetStream\VisionAgents;

use GetStream\Exceptions\StreamApiException;
use GetStream\Exceptions\StreamException;
use GetStream\GeneratedModels\CallRequest;
use GetStream\GeneratedModels\GetOrCreateCallRequest;
use GetStream\VideoClient;
use GetStream\VisionAgents\Edge\StreamTransport;
use GetStream\VisionAgents\Exception\ConfigurationException;
use GetStream\VisionAgents\Exception\RouterException;
use Psr\Http\Client\ClientInterface;

/**
 * The Stream call an agent is asked to join, through Stream's own PHP SDK.
 *
 * The acceleration backend joins a call that already exists, so what is needed here is creating
 * one and a link a person can open to be on the other end of it. No media crosses this class.
 *
 * Server side only, because it holds the app secret.
 */
final readonly class Edge
{
    public const string DEFAULT_CALL_TYPE = 'agent';
    public const string STREAM_API = 'https://chat.stream-io-api.com';
    public const string DEFAULT_MONITOR_URL = 'https://getstream.io/video/demos';

    public VideoClient $video;
    private string $apiKey;
    private string $monitorUrl;

    /**
     * @param ?string $apiKey the Stream app calls are created in; `STREAM_API_KEY`
     * @param ?string $apiSecret its secret, which signs the tokens; `STREAM_API_SECRET`
     * @param string $baseUrl where Stream's API is
     * @param ?string $monitorUrl another deployment of the demo page to link to
     */
    public function __construct(
        ?string $apiKey = null,
        ?string $apiSecret = null,
        string $baseUrl = self::STREAM_API,
        ?string $monitorUrl = null,
        ?ClientInterface $http = null,
    ) {
        $this->apiKey = $apiKey ?? Backend::env(Backend::API_KEY_ENV) ?? '';
        $secret = $apiSecret ?? Backend::env(Backend::API_SECRET_ENV) ?? '';
        if ($this->apiKey === '' || $secret === '') {
            throw new ConfigurationException(Backend::API_KEY_ENV . ' and ' . Backend::API_SECRET_ENV . ' are required to create a call');
        }
        $this->monitorUrl = rtrim($monitorUrl ?? Backend::env('EXAMPLE_BASE_URL') ?? self::DEFAULT_MONITOR_URL, '/');
        $this->video = new VideoClient($this->apiKey, $secret, $baseUrl, new StreamTransport($http ?? Http::client(), Http::requests(), Http::streams()));
    }

    /**
     * Creates the call the backend will join, or returns the one already under that id.
     *
     * An empty id names a new call after a random one, which is what a one-off conversation
     * wants.
     */
    public function createCall(Call $call, string $createdBy): Call
    {
        if ($createdBy === '') {
            throw new ConfigurationException('a call needs somebody to have created it');
        }
        $named = new Call($call->id === '' ? bin2hex(random_bytes(8)) : $call->id, $call->type === '' ? self::DEFAULT_CALL_TYPE : $call->type);
        try {
            $this->video->getOrCreateCall($named->type, $named->id, new GetOrCreateCallRequest(data: new CallRequest(createdByID: $createdBy)));
        } catch (StreamApiException $refused) {
            throw new RouterException($refused->getStatusCode(), "POST /api/v2/video/call/{$named->type}/{$named->id}", $refused->getMessage(), '', 0, $refused);
        } catch (StreamException $failed) {
            throw new RouterException(0, "POST /api/v2/video/call/{$named->type}/{$named->id}", $failed->getMessage(), '', 0, $failed);
        }
        return $named;
    }

    /**
     * A token for somebody to join a call as. Signed here, so this makes no request.
     */
    public function token(string $userId, int $validitySeconds = 3600): string
    {
        if ($userId === '') {
            throw new ConfigurationException('a token needs a user to name');
        }
        return $this->video->createUserToken($userId, [], $validitySeconds);
    }

    /**
     * A link a person can open to join a call from a browser and hear the agent.
     */
    public function monitorUrl(Call $call, string $userId, string $userName = ''): string
    {
        if ($call->id === '') {
            throw new ConfigurationException('there is no call to watch');
        }
        return $this->monitorUrl . '/join/' . rawurlencode($call->id) . '?' . http_build_query([
            'api_key' => $this->apiKey,
            'token' => $this->token($userId),
            'skip_lobby' => 'true',
            'user_name' => $userName === '' ? $userId : $userName,
        ]);
    }
}

File: sdks/dotnet/src/GetStream.VisionAgents/Edge.cs
using System.Security.Cryptography;
using GetStream.Models;

namespace GetStream.VisionAgents;

/// <summary>One Stream call, named the way the backend needs it named.</summary>
public sealed record Call(string Id, string Type = Edge.DefaultCallType);

/// <summary>The Stream credentials calls are created with. Empty fields read the environment.</summary>
public sealed record EdgeOptions
{
    /// <summary>The Stream app's key. Falls back to <c>STREAM_API_KEY</c>.</summary>
    public string? ApiKey { get; init; }

    /// <summary>Its secret, which signs the tokens. Falls back to <c>STREAM_API_SECRET</c>.</summary>
    public string? ApiSecret { get; init; }

    /// <summary>Stream's API, for pointing at another deployment.</summary>
    public string? BaseUrl { get; init; }

    /// <summary>
    /// Points the monitoring link at another deployment of the demo. Falls back to
    /// <c>EXAMPLE_BASE_URL</c>, then Stream's hosted one.
    /// </summary>
    public string? MonitorUrl { get; init; }

    /// <summary>The <see cref="HttpClient"/> Stream's SDK sends with, for <c>IHttpClientFactory</c>.</summary>
    public HttpClient? HttpClient { get; init; }
}

/// <summary>
/// Creates the Stream calls an agent joins, and mints the tokens for watching them.
/// </summary>
/// <remarks>
/// The router joins a call; it does not create one. Creating it is Stream's own API, which
/// Stream's server-side SDK (<c>getstream-net</c>) already speaks, so this is that SDK
/// rather than a second copy of it.
/// </remarks>
public sealed class Edge
{
    /// <summary>The Stream call type used when none is named.</summary>
    public const string DefaultCallType = "agent";

    /// <summary>Stream's hosted video demo, the page the Python examples open.</summary>
    public const string DefaultMonitorUrl = "https://getstream.io/video/demos";

    // A call outliving the browser's token is a call nobody is still on.
    private static readonly TimeSpan MonitorTokenValidity = TimeSpan.FromHours(1);

    private readonly StreamClient _stream;
    private readonly VideoClient _video;
    private readonly string _monitorUrl;

    /// <summary>Reads whatever the options leave empty out of the environment.</summary>
    /// <exception cref="ConfigurationException">There is no key or no secret.</exception>
    public Edge(EdgeOptions? options = null)
    {
        options ??= new EdgeOptions();
        ApiKey = options.ApiKey ?? Backend.Env("STREAM_API_KEY") ?? "";
        var secret = options.ApiSecret ?? Backend.Env("STREAM_API_SECRET") ?? "";
        if (ApiKey == "" || secret == "")
        {
            throw new ConfigurationException("STREAM_API_KEY and STREAM_API_SECRET are required to create a call");
        }
        var streamOptions = new StreamOptions { ApiKey = ApiKey, ApiSecret = secret };
        if (options.BaseUrl is { Length: > 0 } baseUrl)
        {
            streamOptions.BaseUrl = baseUrl;
        }
        if (options.HttpClient is { } http)
        {
            streamOptions.HttpClient = http;
        }
        _stream = new StreamClient(streamOptions);
        _video = new VideoClient(_stream);
        _monitorUrl = (options.MonitorUrl ?? Backend.Env("EXAMPLE_BASE_URL") ?? DefaultMonitorUrl).TrimEnd('/');
    }

    /// <summary>The Stream app's key.</summary>
    public string ApiKey { get; }

    /// <summary>
    /// Creates the call the backend will join, or returns the one already under that id.
    /// </summary>
    /// <remarks>A null id names a new call after a random one, which is what a one-off conversation wants.</remarks>
    public async Task<Call> CreateCallAsync(string? id, string? type, string createdBy, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(createdBy))
        {
            throw new ConfigurationException("a call needs somebody to have created it");
        }
        var call = new Call(string.IsNullOrEmpty(id) ? Convert.ToHexStringLower(RandomNumberGenerator.GetBytes(8)) : id,
            string.IsNullOrEmpty(type) ? DefaultCallType : type);
        try
        {
            await _video.GetOrCreateCallAsync(call.Type, call.Id,
                new GetOrCreateCallRequest { Data = new CallRequest { CreatedByID = createdBy } },
                cancellationToken).ConfigureAwait(false);
        }
        catch (GetStreamException failure)
        {
            throw new RouterException(failure is GetStreamApiException api ? api.StatusCode : 0,
                $"POST /video/call/{call.Type}/{call.Id}", failure.Message, failure);
        }
        return call;
    }

    /// <summary>Mints a token for somebody to join a call as. Signed here, so no request is made.</summary>
    public string Token(string userId, TimeSpan? validity = null) =>
        string.IsNullOrEmpty(userId)
            ? throw new ConfigurationException("a token needs a user to name")
            : _stream.CreateUserToken(userId, validity ?? MonitorTokenValidity);

    /// <summary>A link a person can open to join a call from a browser and hear the agent.</summary>
    public string MonitorUrl(Call call, string userId, string? userName = null)
    {
        if (string.IsNullOrEmpty(call.Id))
        {
            throw new ConfigurationException("there is no call to watch");
        }
        var query = string.Join('&',
            $"api_key={Uri.EscapeDataString(ApiKey)}",
            $"token={Uri.EscapeDataString(Token(userId))}",
            "skip_lobby=true",
            $"user_name={Uri.EscapeDataString(string.IsNullOrEmpty(userName) ? userId : userName)}");
        return $"{_monitorUrl}/join/{Uri.EscapeDataString(call.Id)}?{query}";
    }
}
